src/main/java/net/minecraftforge/common/DimensionDictionary.java

```diff
@@ -0,0 +1,313 @@
+    }
+
+    /**
+     * Registers many dimension types.
+     *
+     * @param types The types to register.
+     */
+    public static void registerTypes(Collection<String> types)
+    {
+        for (String type : types)
+            registerType(type);
+    }
+
+    /**
+     * Adds subtypes to a specific type.
+     * This will also register any used types, if they weren't already.
+     *
+     * @param type     The type.
+     * @param subtypes The subtypes.
+     */
+    public static void registerSubtypes(String type, Collection<String> subtypes)
+    {
+        registerGetType(type).subtypes.addAll(subtypes);
+
+        for (String sub : subtypes)
+            registerGetType(sub).supertypes.add(type);
+    }
+
+    /**
+     * Adds supertypes to a specific type.
+     * This will also register any used types, if they weren't already.
+     *
+     * @param type     The type.
+     * @param supertypes The supertypes.
+     */
+    public static void registerSupertypes(String type, Collection<String> supertypes)
+    {
+        registerGetType(type).supertypes.addAll(supertypes);
+
+        for (String supertype : supertypes)
+            registerGetType(supertype).subtypes.add(type);
+    }
+
+    /**
+     * Returns a set of types a specific dimension was registered for.
+     *
+     * @param provider The dimension's provider.
+     * @return The dimension's types.
+     */
+    public static Set<String> getDimensionTypes(WorldProvider provider)
+    {
+        if (provider instanceof Handler)
+            return ((Handler) provider).getDimensionTypes();
+            
+        Set<String> types = dimensionTypes.get(provider.dimensionId);
+        return types != null ? SET_UNCATEGORIZED : Collections.<String>emptySet();
+    }
+
+    /**
+     * Determines if a dimension matches all types from a collection of types.
+     *
+     * @param provider The dimension's provider.
+     * @param types       The types.
+     * @return True if all types were matched, otherwise false.
+     */
+    public static boolean dimensionMatchesAllTypes(WorldProvider provider, Collection<String> types)
+    {
+        for (String type : types)
+        {
+            if (!dimensionMatchesType(provider, type))
+                return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Determines if a dimension matches a specific type.
+     * This is the case exactly when the dimension is associated with either the type or any of its subtypes.
+     *
+     * @param provider The dimension's provider.
+     * @param type        The type.
+     * @return True if the dimension matches the type, otherwise false.
+     */
+    public static boolean dimensionMatchesType(WorldProvider provider, String type)
+    {
+        Set<String> dimTypes = getDimensionTypes(provider);
+
+        Queue<String> curTypes = new ArrayDeque<String>();
+        curTypes.add(type);
+
+        String curType;
+
+        while ((curType = curTypes.poll()) != null)
```

Comment:
Why is the type string put into a collection only to pull it out of the collection?